### PR TITLE
Refactor time util and improve diagnostic logging with timezone friendly date and time for Brave Ads events

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/9199

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

- Confirm "Next payment date" on ads panel is working as expected

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_ads/browser/ads_service_impl.cc [hunk 1]
```diff
@@ -1618,10 +1619,11 @@ void AdsServiceImpl::StartRemoveOnboardingTimer() {
       base::TimeDelta::FromSeconds(timer_offset_in_seconds),
           base::BindOnce(&AdsServiceImpl::RemoveOnboarding, AsWeakPtr()));
 
-  auto time = base::TimeFormatFriendlyDateAndTime(
-      base::Time::FromDoubleT(timestamp_in_seconds));
+  const std::string friendly_date_and_time =
+      base::UTF16ToUTF8(base::TimeFormatFriendlyDateAndTime(
+          base::Time::FromDoubleT(now_in_seconds + timer_offset_in_seconds)));
```
Comment (zenparsing) [inline components/brave_ads/browser/ads_service_impl.cc:1624]: Isn't `now_in_seconds + timer_offset_in_seconds` equal to `timestamp_in_seconds`?

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/per_hour_frequency_cap.cc [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
 #include "bat/ads/internal/frequency_capping/exclusion_rules/per_hour_frequency_cap.h"
 #include "bat/ads/internal/frequency_capping/frequency_capping.h"
-#include "bat/ads/internal/time.h"
+#include "bat/ads/internal/time_util.h"
```
Comment (zenparsing) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/per_hour_frequency_cap.cc:8]: We don't seem to be using anything defined in `time_util.h`, is that right? Are we including it just to get `base/time/time.h`? I see a bunch of similar includes of `time_util.h`.

File: vendor/bat-native-ads/src/bat/ads/internal/reports.cc [hunk 3]
```diff
@@ -31,7 +32,8 @@ std::string Reports::GenerateAdNotificationEventReport(
   rapidjson::StringBuffer buffer;
   rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
 
-  const std::string timestamp = Time::Timestamp();
+  const base::Time time = base::Time::Now();
+  const std::string timestamp = FriendlyDateAndTime(time, false);
```
Comment (zenparsing) [inline vendor/bat-native-ads/src/bat/ads/internal/reports.cc:36]: I suppose this isn't really a "timestamp" anymore, but rather a formatted date string, but it looks like renaming the variable/JSON field would be painful.

File: vendor/bat-native-confirmations/src/bat/confirmations/internal/confirmations_impl.cc [hunk 4]
```diff
@@ -1040,8 +1040,13 @@ uint64_t ConfirmationsImpl::GetAdNotificationsReceivedThisMonthForTransactions(
   now.UTCExplode(&now_exploded);
 
   for (const auto& transaction : transactions) {
+    if (transaction.timestamp_in_seconds == 0) {
+      // Workaround for Windows crash when passing 0 to UTCExplode
```
Comment (zenparsing) [inline vendor/bat-native-confirmations/src/bat/confirmations/internal/confirmations_impl.cc:1044]: 👍

File: vendor/bat-native-confirmations/src/bat/confirmations/internal/time_util.cc [hunk 5]
```diff
@@ -0,0 +1,76 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
```
Comment (zenparsing) [inline vendor/bat-native-confirmations/src/bat/confirmations/internal/time_util.cc:1]: It's a shame that we have to duplicate, but 🤷

File: components/brave_ads/browser/ads_service_impl.cc (hunk 1 above)
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc:1624]: In this situation, it is not as if the timestamp has expired we trigger the timer after 1 minute

File: vendor/bat-native-confirmations/src/bat/confirmations/internal/time_util.cc (hunk 5 above)
Comment (tmancey) [reply vendor/bat-native-confirmations/src/bat/confirmations/internal/time_util.cc:1]: Totally agree, and I have raised https://github.com/brave/brave-browser/issues/9611

File: vendor/bat-native-ads/src/bat/ads/internal/reports.cc (hunk 3 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/reports.cc:36]: Logs are being refactored as part of https://github.com/brave/brave-browser/issues/6845 and https://github.com/brave/brave-browser/issues/9533 so reports will be deprecated

**Comment by tmancey** [comment]
CI failed for iOS, macOS and Windows. Restarting

**Comment by tmancey** [comment]
CI failed for macOS and Windows. Restarting

**Comment by tmancey** [comment]
CI failed macOS, restarting macOS

**Comment by tmancey** [comment]
macOS failed on CI, restarting macOS

**Comment by tmancey** [comment]
Failed CI due to known iOS issue and rewards browser tests

**Pull request merged**